### DISCUSSION TITLE
Excess Production in Graph

### DISCUSSION THREAD

**Discussion by nullgel** [jasonacox/Powerwall-Dashboard#176] [Q&A]
A number that I'm obsessed with watching is the (hopefully positive) number denoting the amount of excess energy that I produce. I generate this number this way:

![image](https://user-images.githubusercontent.com/27741277/217572539-ab6bd6d9-ee9f-44c2-b06f-07a857ec2eb5.png)

However, I would really like to watch this number over time in the year long graph:

![image](https://user-images.githubusercontent.com/27741277/217572049-37df50e3-a3d3-496a-ae43-1c5161799651.png)

(Note: In the above graph, the bright yellow line is a 30-day average of solar production and the bright blue is home usage)

But I am having a tough time getting it to work. I feel like I'm over-complicating it and would really appreciate some advice. Even a point in the right direction may be sufficient.

Thanks!

**Comment by jasonacox** [comment]
Hi @nullgel - I love the look of your yearly graph with those moving averages.  Can you share how you did that?

For getting a graph to show the ratio of solar production over home usage, you could use a query like this:

```sql
SELECT sum("solar")/sum("home") 
    FROM "kwh"."http" 
    WHERE $timeFilter GROUP BY time(1d) fill(0) tz('America/Los_Angeles')
```

You would want to add a series override to move it to a separate axis so it shows up.  Here is my test:

<img width="940" alt="image" src="https://user-images.githubusercontent.com/836718/217736269-1092f266-6010-438c-b319-22c0115fa18f.png">

**Comment by nullgel** [comment]
The moving average was nothing special:
![image](https://user-images.githubusercontent.com/27741277/217822679-7c5768a9-8191-4065-a7b6-f3cd0f30d04f.png)

Thank you for taking the time to generate that query, but (as usual) I don't think that I explained myself quite well enough. What I'm ultimately after, is knowing whether or not I'm getting close to owing money to my utility. My utility is a flat rate all day so I'm keeping an eye on how much energy I've banked up with them verses how much I'm importing from them. The number I'm generating instantaneously is pretty accurate, but I want to see its trajectory. It has to be a cumulative calculation for the entire year up to each day.

Can you decipher what I'm getting at? Honestly, it may end up being too processor intensive to do what I want.

**Reply by jasonacox** [reply]
Thanks @nullgel - I love the simple moving-average graph you built.  Looks nice.

Ok, I think I understand what you are looking for.  You are right, it is more complex.  You want data points on the time series graph (t=0 to t=n) to be a sum of the (home/solar) for that duration.  Assuming the graph you want is represented by function `ratio(n)`.  And `home(n)` and `solar(n)` are the sums of home usage for the day `n`.

    ratio(0) = (home(0)/solar(0))
    ratio(1) = (home(0) + home(1)) / (solar(0) + solar(1))
    ratio(n) = (home(0) + ... + home(n)) / (solar(0) + ... + solar(n)) 

Is that right?  If so, we can look at how we reduce or use that to plot.

**Comment by nullgel** [comment]
Exactly! Yeah... I kicked it around for a few more hours yesterday with no
luck. It sure seemed simple at first glance, but it's tricky to implement.

On Fri, Feb 10, 2023 at 1:10 AM Jason Cox ***@***.***> wrote:

> Thanks @nullgel <https://github.com/nullgel> - I love the simple
> moving-average graph you built. Looks nice.
>
> Ok, I think I understand what you are looking for. You are right, it is
> more complex. You want data points on the time series graph (t=0 to t=n) to
> be a sum of the (home/solar) for that duration. Assuming the graph you want
> is represented by function ratio(n). And home(n) and solar(n) are the
> sums of home usage for the day n.
>
> ratio(0) = (home(0)/solar(0))
> ratio(1) = (home(0) + home(1)) / (solar(0) + solar(1))
> ratio(n) = (home(0) + ... + home(n)) / (solar(0) + ... + solar(n))
>
> Is that right? If so, we can look at how we reduce or use that to plot.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/jasonacox/Powerwall-Dashboard/discussions/176#discussioncomment-4928596>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AGTUYXMFZ6WCUARU2BZRFE3WWXLVRANCNFSM6AAAAAAUVM35UE>
> .
> You are receiving this because you were mentioned.Message ID:
> <jasonacox/Powerwall-Dashboard/repo-discussions/176/comments/4928596@
> github.com>
>


-- 
Nate Winesett

**Reply by youzer-name** [reply]
Those who've read a few of my posts won't be surprised that I went to MySQL to solve this... but I was able to do this using the KWH data that I copy to MySQL. It required the use of a CTE (common table expression) query. I can't write queries like this from scratch, I had to adapt it from something I found online that was showing how to do cumulative totals over time.

 I have no idea if anything like this is possible in InfluxDB. It definitely can't be done by the same method, as there is nothing like a CTE query in InfluxDB 1.8 AFAIK.

I can add the details of how I built this to one of my posts about using MySQL if there's interest.

![image](https://user-images.githubusercontent.com/57046429/218194302-1f7e3c77-e52c-4ef1-b3a1-ef54bef0fc5a.png)

The yellow line is cumulative solar production. 
The blue line is cumulative home. 
The filled area is on the right Y axis and is ((solar / home) * 100). It would fill green if it passed 100%, and is light blue when < 100%

I only got my powerwalls at the end of February 2022, so the year's data starts there. This year so far looks like this:

![image](https://user-images.githubusercontent.com/57046429/218194537-0612dab1-98e0-4d75-a48f-4b16636fd402.png)

I didn't originally read carefully enough, so the first thing I came up with was calculating the percentage on a daily basis, rather than annual, so it looked like this:

![image](https://user-images.githubusercontent.com/57046429/218189740-15a96cdc-8710-44f5-94e0-b3bceb7fe774.png)

Here you can see the color fill changing when above 100%. At mid-day on the second day I had produced solar equal to 250% of that day's home use, and by the end of the day it was 138%. On the first day I only got slightly ahead on solar for a bit in the middle of the day, and ended the day having produced 76% of the home's use by solar.

**Reply by jasonacox** [reply]
@youzer-name - these are amazing!  I think @nullgel was looking for the inverse (home/solar) but I like this look too!
